### PR TITLE
Add span stacks

### PR DESCRIPTION
**Pull request by bwoebi**
### Description

We need some form of span stacks for fibers, to be able to attach active spans and their ancestors to a specific fiber and switch these out along any fiber switches.

Starting out with designing tests to figure out the semantics. 

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [ ] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

**Review by pierotibou — COMMENTED** [review]
A few set of comments

File: tests/ext/span_stack/span_stack_swap_traced_function.phpt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    DDTrace\switch_stack($a_span_on_another_stack);
+    DDTrace\close_span();
+    echo 'Now, we have explicitly closed it: '; var_dump($a_span_on_another_stack->getDuration() !== 0);
+}
+
+function creates_span_stack() {
+    $new_stack = DDTrace\create_stack();
+    $inner_span = inner();
+    echo 'The stack stays intact within the function: '; var_dump($new_stack == DDTrace\active_stack());
+    echo 'And the span started within the inner function was automatically closed, given that it is on the same span stack: '; var_dump($inner_span->getDuration() === 0);
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply tests/ext/span_stack/span_stack_swap_traced_function.phpt:26]: traced functions cause the associated spans (and all opened spans higher up on the stack) to be automatically closed. (that's how it is currently)

[the remark this answers is not in the document]
Comment (bwoebi) [reply tests/ext/span_stack/span_stack_swap_traced_function.phpt:26]: And yes, that's a mistake in the test, it should be !== null.

File: tests/ext/span_stack/span_trace_swap.phpt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+$new_root = DDTrace\start_trace_span();
+
+DDTrace\switch_stack($primary_trace);
+echo 'We are back on our primary stack: '; var_dump($primary_trace == DDTrace\active_span());
+
+# Closes the primary trace itself
+DDTrace\close_span();
+
+echo 'We closed the primary trace. No other span is active right now: '; var_dump(null == DDTrace\active_span());
+echo 'And no other stack is active either: '; var_dump(null == DDTrace\active_stack());
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply tests/ext/span_stack/span_trace_swap.phpt:21]: To not leak a stack, we automatically close stacks which do not hold any references. (see the span_trace_stack_autoclose.phpt test)

File: tests/ext/span_stack/start_span_new_trace.phpt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+echo 'New trace span is reflected in DDTrace\active_stack(): '; var_dump($new_root->stack == DDTrace\active_stack());
+echo 'New trace span stack has no parent: '; var_dump($new_root->stack->parent == null);
+echo 'New trace span has no parent: '; var_dump($new_root->parent == null);
+echo 'New trace span has a trace id equal to itself: '; var_dump($new_root->id == DDTrace\trace_id());
+
+# on new root
+$active_new_root_span = DDTrace\start_span();
+
+DDTrace\switch_stack($primary_trace);
+echo 'stack update successful: '; var_dump($primary_trace == DDTrace\stack_span());
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply tests/ext/span_stack/start_span_new_trace.phpt:25]: yep, that's an artifact from the earlier iteration.

File: tests/ext/span_stack/start_top_level_span_stack.phpt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+DDTrace\close_span();
+echo 'Verify the stack_span stays if the top-level span is closed - this span stack is not tied to a trace directly: '; var_dump($new_stack == DDTrace\active_stack());
+print 'There is no active span now: '; var_dump(DDTrace\active_span() == null);
+
+# Attempt closing on an empty span stack
+DDTrace\close_span();
+echo 'Given no active span, the active span stays null: '; var_dump(DDTrace\active_span() == null);
+echo 'This also must not affect the active span stack: '; var_dump($new_stack == DDTrace\active_stack());
+
+DDTrace\switch_stack();
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply tests/ext/span_stack/start_top_level_span_stack.phpt:27]: In general, our API has been designed to be forgiving, like close_span() without a span existing does nothing. This is just continuing the current design philosophy.

**Review by krakjoe — APPROVED** [review]

**Pull request merged**